### ISSUE TITLE
Dynamic Permissions Not Updating in AdministrationService Without Rebuild

### ISSUE THREAD

**Issue by alihdev** [abpframework/eShopOnAbp#214]
Hi,

We need to implement dynamic permissions to add permissions at runtime. Currently, we're using **IPermissionDefinitionRecordRepository** to add new permissions and **IDynamicPermissionDefinitionStoreInMemoryCache** to clear the store.

When we add a new permission from service X at runtime, the addition is successful. The permission gets added to the **AbpPermissions** table in the AdministrationService.

However, the issue arises when we call the application configuration endpoint; to get permissions .. the newly added permissions don't appear unless we rebuild the AdministrationService.

Steps to Reproduce:

- Enable IsDynamicPermissionStoreEnabled in the AdministrationServiceHttpApiHostModule.
```c#
Configure<PermissionManagementOptions>(options =>
{
    options.IsDynamicPermissionStoreEnabled = true;
});
```
- Make changes to permissions in another microservice.
using IPermissionDefinitionRecordRepository & IDynamicPermissionDefinitionStoreInMemoryCache
- Expectation: Permissions should dynamically update in the Admin Service without requiring a rebuild.
- Actual result: Permissions only update after rebuilding the Admin Service.

Note: In a monolithic application setup, this dynamic permission addition works seamlessly without encountering any issues

**Comment by maliming** [comment]
> However, the issue arises when we call the application configuration endpoint; to get permissions .. the newly added permissions don't appear unless we rebuild the AdministrationService.

You can wait for 30 seconds to refresh the memory cache.

https://github.com/abpframework/abp/blob/dev/modules/permission-management/src/Volo.Abp.PermissionManagement.Domain/Volo/Abp/PermissionManagement/DynamicPermissionDefinitionStore.cs#L92

**Comment by alihdev** [comment]
Hello,

Thank you for your prompt response. the refresh isn't happening even after waiting for more than 30 seconds.

Regarding our code to clear the cache:

```c#
protected virtual Task ClearStoreCacheAsync()
{
    _storeCache.LastCheckTime = null;
    _storeCache.CacheStamp = null;

    return Task.CompletedTask;
}
```

We're calling this code after inserting new permissions.

Do you have any further insights into why the cache might not be refreshing as expected?

Thank you.

**Comment by maliming** [comment]
hi

Can you share a project to show this problem? You can upload it to GitHub.

**Comment by alihdev** [comment]
We'll work on creating a project that highlights the issue

**Comment by alihdev** [comment]
Hello,

While debugging the ABP code, we've discovered that when the cache is cleared or a role is assigned to a user from the X service, the admin service isn't affected. However, when we clear the cache from the admin service, everything works fine.

**Please refer to this video** for a clearer understanding of the issue: [Video Link](https://github.com/alihdev/DotnetAbp.Issues/blob/eShop-dynamicPrmissions/Video_2023-12-23_150704.wmv)

**Additionally**, we've created a project for your review to investigate the problem: [Project Link](https://github.com/alihdev/DotnetAbp.Issues/tree/eShop-dynamicPrmissions)

**Please note** the changes we've made in the latest version of eShop code:

- Permission creation from CmskitService: [Code Link](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/cmskit/src/EShopOnAbp.CmskitService.Application/Samples/SampleAppService.cs#L43)

- Dynamic permission enablement in AdministrationService: [Code Link](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/administration/src/EShopOnAbp.AdministrationService.HttpApi.Host/AdministrationServiceHttpApiHostModule.cs#L69)


Your assistance in resolving this matter is greatly appreciated.

Thank you!

**Comment by maliming** [comment]
hi

Can you update your `tye.yaml` file and share the steps to reproduce?

I have some doubts about running your app and authentication.

**Comment by alihdev** [comment]
The project I shared with you is the same as [eShopOnAbp](https://github.com/abpframework/eShopOnAbp/tree/main). I downloaded it, and we've made the following changes:

> **Please note** the changes we've made in the latest version of eShop code:
> 
> * Permission creation from CmskitService: [Code Link](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/cmskit/src/EShopOnAbp.CmskitService.Application/Samples/SampleAppService.cs#L43)
> * Dynamic permission enablement in AdministrationService: [Code Link](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/administration/src/EShopOnAbp.AdministrationService.HttpApi.Host/AdministrationServiceHttpApiHostModule.cs#L69)
> 

Additionally, we modified the [type.yml](https://github.com/alihdev/DotnetAbp.Issues/blob/eShop-dynamicPrmissions/tye.yaml) to only run the **cmskit-service** and **administration-service**, similar to the setup in [tye.yaml](https://github.com/abpframework/eShopOnAbp/blob/main/tye.yaml). You can use it.

**Comment by maliming** [comment]
I will run the `run-tye.ps1`,  What are the next steps?

**Comment by alihdev** [comment]
When the gateway is open, please open two tabs: one for the **administration-service** and another for the **cmskit-service**. 

In the **cmskit-service** Swagger, call the endpoint [CreatePermission](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/cmskit/src/EShopOnAbp.CmskitService.Application/Samples/SampleAppService.cs#L43) in SimpleAppService to create dynamic permissions. 
Subsequently, 
In **administration-service** Swagger, call the endpoint **Get** in AbpApplicationConfigurationAppService to retrieve the current user with their permissions.

You'll notice that the permission (created in [CreatePermission](https://github.com/alihdev/DotnetAbp.Issues/blob/14e554c940c0cb787d69e29e2e92fa894b563f49/services/cmskit/src/EShopOnAbp.CmskitService.Application/Samples/SampleAppService.cs#L43)) is not found only after you rebuild the admin service. 

Further details can be found in the [commmet](https://github.com/abpframework/eShopOnAbp/issues/214#issuecomment-1868449962)

**Comment by maliming** [comment]
hi

You can inject the `IDistributedCache DistributedCache` and remove the `$"{CacheOptions.KeyPrefix}_AbpInMemoryPermissionCacheStamp";` cache.

**Comment by alihdev** [comment]
Ok, I tried this, and the cache cleared successfully. 

However, I encountered another issue when attempting to call this function from X service:

```
  await _permissionManager.SetForRoleAsync("admin", name, true);
```
We've discovered that we should also enable the [IsDynamicPermissionStoreEnabled](https://github.com/abpframework/abp/blob/054bad8914382a1c2ef9c672d631114b591473ff/modules/permission-management/src/Volo.Abp.PermissionManagement.Domain/Volo/Abp/PermissionManagement/DynamicPermissionDefinitionStore.cs#L77C14-L77C41) configuration in the X service when we activate it. We're encountering the following exceptions within the [PermissionManager](https://github.com/abpframework/abp/blob/054bad8914382a1c2ef9c672d631114b591473ff/modules/permission-management/src/Volo.Abp.PermissionManagement.Domain/Volo/Abp/PermissionManagement/PermissionManager.cs#L239)

```
ManagementProviders 
(Autofac.Core.DependencyResolutionException: An exception was thrown while activating Volo.Abp.PermissionManagement.Identity.RolePermissionManagementProvider. ---> Autofac.Core.DependencyResolutionException: None of the constructors found on type 'Volo.A…):
```

```
[humanresource-service_3e0ba3d1-e]: [13:39:17 WRN] An exception occurred, but response has already started!
[humanresource-service_3e0ba3d1-e]: [13:39:17 ERR] An unhandled exception has occurred while executing the request.
[humanresource-service_3e0ba3d1-e]: Autofac.Core.DependencyResolutionException: An exception was thrown while activating Volo.Abp.PermissionManagement.Identity.RolePermissionManagementProvider.
[humanresource-service_3e0ba3d1-e]: ---> Autofac.Core.DependencyResolutionException: None of the constructors found on type 'Volo.Abp.PermissionManagement.Identity.RolePermissionManagementProvider' can be invoked with the available services and parameters:
[humanresource-service_3e0ba3d1-e]: Cannot resolve parameter 'Volo.Abp.Identity.IUserRoleFinder userRoleFinder' of constructor 'Void .ctor(Volo.Abp.PermissionManagement.IPermissionGrantRepository, Volo.Abp.Guids.IGuidGenerator, Volo.Abp.MultiTenancy.ICurrentTenant, Volo.Abp.Identity.IUserRoleFinder)'.
[humanresource-service_3e0ba3d1-e]:
[humanresource-service_3e0ba3d1-e]: See https://autofac.rtfd.io/help/no-constructors-bindable for more info.
[humanresource-service_3e0ba3d1-e]: at Autofac.Core.Activators.Reflection.ReflectionActivator.<>c__DisplayClass14_0.<UseSingleConstructorActivation>b__0(ResolveRequestContext ctxt, Action`1 next)
[humanresource-service_3e0ba3d1-e]: at Autofac.Core.Resolving.Middleware.DisposalTrackingMiddleware.Execute(ResolveRequestContext context, Action`1 next)
[humanresource-service_3e0ba3d1-e]: at Autofac.Builder.RegistrationBuilder`3.<>c__DisplayClass41_0.<PropertiesAutowired>b__0(ResolveRequestContext ctxt, Action`1 next)
[humanresource-service_3e0ba3d1-e]: at Autofac.Core.Resolving.Middleware.ActivatorErrorHandlingMiddleware.Execute(ResolveRequestContext context, Action`1 next)
```

Also, here's an image related to the code issue: 

![image](https://github.com/abpframework/eShopOnAbp/assets/97635301/480ee99d-8960-4878-8676-d441ee1ebf30)

I will reproduce this issue in the public [project](https://github.com/alihdev/DotnetAbp.Issues/tree/eShop-dynamicPrmissions)

Thank you!

**Comment by alihdev** [comment]
Related to the above [comment](https://github.com/abpframework/eShopOnAbp/issues/214#issuecomment-1870849504), I have fixed it; there was a dependency issue in the host (AbpIdentityHttpApiClientModule).

**Comment by maliming** [comment]
https://github.com/abpframework/abp/issues/18628

**Comment by maliming** [comment]
> hi
> 
> You can inject the `IDistributedCache DistributedCache` and remove the `$"{CacheOptions.KeyPrefix}_AbpInMemoryPermissionCacheStamp";` cache.
> 
> I will find a way on the framework.

hi

I checked the source code.

The `StaticPermissionSaver` will update the cache value.

https://github.com/abpframework/abp/blob/dev/modules/permission-management/src/Volo.Abp.PermissionManagement.Domain/Volo/Abp/PermissionManagement/StaticPermissionSaver.cs#L116-L123

**Comment by maliming** [comment]
hi @StevenNgoc 

Please create a new issue, and share the steps.

Thanks.
